cfme/tests/cloud_infra_common/test_provisioning.py

```diff
@@ -124,33 +123,57 @@ def test_gce_preemptible_provision(appliance, provider, instance_args, soft_asse
     soft_assert(instance.exists_on_provider, "Instance wasn't provisioned successfully")
 
 
-@pytest.mark.manual
-@pytest.mark.meta(coverage=[1676910])
-def test_provision_approval_deny():
-    """ Test whether the approver can deny the request.
+def post_approval(smtp_test, provision_request, vm_type, requester, provider, vm_names):
+    # requester includes the trailing space
+    approved_subject = normalize_text("your {} request was approved".format(vm_type))
+    approved_from = normalize_text("{} request from {}was approved".format(vm_type, requester))
 
-    Prerequisities:
-        * A provider that can provision.
+    wait_for_messages_with_subjects(smtp_test, [approved_subject, approved_from], num_sec=90)
 
-    Steps:
-        * Create a provisioning request that does not get automatically approved (eg. ``num_vms``
-            bigger than 1)
-        * Deny the created request
+    smtp_test.clear_database()
 
-    Polarion:
-        assignee: jhenner
-        caseimportance: high
-        casecomponent: Provisioning
-        initialEstimate: 1/3h
+    # Wait for the VM to appear on the provider backend before proceeding to ensure proper cleanup
+    logger.info('Waiting for vms %s to appear on provider %s', ", ".join(vm_names), provider.key)
+    wait_for(
+        lambda: all(map(provider.mgmt.does_vm_exist, vm_names)),
+        handle_exception=True,
+        num_sec=600
+    )
+
+    provision_request.wait_for_request(method='ui')
+    msg = "Provisioning failed with the message {}".format(provision_request.row.last_message.text)
+    assert provision_request.is_succeeded(method='ui'), msg
+
+    # account for multiple vms, specific names
+    completed_subjects = [
+        normalize_text("your {} request has completed vm name {}".format(vm_type, name))
+        for name in vm_names
+    ]
+    wait_for_messages_with_subjects(smtp_test, completed_subjects, num_sec=90)
+
+
+def wait_for_messages_with_subjects(smtp_test, expected_subjects, num_sec):
+    """ This waits for all the expected subjects to be present the list of received
+    mails with partial match.
     """
-    pass
+    def check_subjects():
+        subjects = [normalize_text(m["subject"]) for m in smtp_test.get_emails()]
+        # Looking for each expected subject in the list of received subjects with partial match
+        for subject in expected_subjects:
+            if not any(subject in r_sub for r_sub in subjects):
+                logger.info('Subject {subjects} not found in the received emails yet.')
+                return False
+        logger.info('All expected subjects found in the received email list.')
+        return True
```

Comment:
Q: Could you do?
```suggestion
        return subjects in expected_subjects
```

Reply:
From what you have now, it looks like `subjects` and `expected_subjects` are not exactly of the same structure, is there someway you could enforce that to simplify the logic of this method?

Reply:
> From what you have now, it looks like `subjects` and `expected_subjects` are not exactly of the same structure, is there someway you could enforce that to simplify the logic of this method?

I was thinking about the same... the code looks strange as it is and I though why it is not written the way as you suggest. I think it is for reason is that we need to check the `subject` to be just a substring of the strings in `expected_subjects`.

Perhaps it is something worthy to check and comment on it in the code or change the code for this to be clear.

Note I also discovered that the `smtp_test` "mock" provides a `subject_like` argument, but I figured out I cannot use it and keep the behaviour same enough for my liking.

Reply:
I see, I think a comment explaining this will suffice for now. Then we can evaluate whether to cleanup this logic a bit later.